### PR TITLE
Use KubeletServer.NodeIP instead of KubeletServer.HostnameOverride to set node IP

### PR DESCRIPTION
**Pull request by pravisankar**

### CODE REVIEW

**Comment by pravisankar** [comment]
Depends on https://github.com/kubernetes/kubernetes/pull/18541

**Comment by liggitt** [comment]
Changing back could trigger more attempts by the node to delete and recreate its API object (which prevents the node from starting)

**Comment by liggitt** [comment]
Need to resolve https://github.com/kubernetes/kubernetes/issues/17731 before this can merge

**Comment by pravisankar** [comment]
[test]

**Comment by pravisankar** [comment]
@liggitt @DirectXMan12 @ncdc @eparis PTAL

**Comment by liggitt** [comment]
change `UPSTREAM: <drop>: ...` to `UPSTREAM: 18541: ...` if there's an existing PR

**Comment by liggitt** [comment]
can you outline the ways you tested this? Off the top of my head, I'd expect:

- upgrade path with nodeIP set
  - launch 1.0, with nodeIP set in config
  - relaunch 1.1, with nodeIP set in config, against the same etcd
  - relaunch with this PR, with nodeIP set in config, against the same etcd
  - expect node to come up, with warning, and pods to be able to be deployed

- upgrade path without nodeIP set
  - launch 1.0, without nodeIP set in config
  - relaunch 1.1, without nodeIP set in config, against the same etcd
  - relaunch with this PR, without nodeIP set in config, against the same etcd
  - expect node to come up, with no warning
  - relaunch with this PR, with nodeIP set in config, against the same etcd
  - expect node to come up, with no warning, and pods to be able to be deployed

**Comment by DirectXMan12** [comment]
For the part that I wrote:

Before my patch:

1. launch node with nodeIP set (giving `externalID == nodeIP`)
2. relaunch node without nodeIP set (giving `externalID == hostname`)
3. Fail

After my patch:


1. Try to launch node again
2. It works, gives warning about `externalID` mismatch
3. Delete node
4. Launch node with same config (`externalID == hostname`)
5. Relaunch node with nodeIP set (`externalID == nodeIP`)
6. It works, gives warning about externalID mismatch

**Comment by pravisankar** [comment]
Updated the PR:
Changed commit message: UPSTREAM: <drop>: ... to UPSTREAM: 18541: ..

For tolerating node externalID to have hostname or node IP in case of non-cloud provider, we just need to look at the currentNode.Spec.ExternalID (instead of node.Spec.ExternalID). 
@DirectXMan12  please review?

Tested various combinations from v1.0.8 to v1.1.1 without fix and with fix. I think, I covered all the test cases mentioned by @liggitt and @DirectXMan12 

@liggitt @ncdc @eparis please review/merge

**Comment by DirectXMan12** [comment]
@pravisankar LGTM.  The logic you removed was for the "hostname -> nodeIP" case, but that should be solved by your other changes, correct?

**Comment by liggitt** [comment]
LGTM

**Comment by eparis** [comment]
LGTM   who is going to click the button?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -945,6 +989,22 @@ func (kl *Kubelet) registerWithApiserver() {
 				kl.registrationCompleted = true
 				return
 			}
+
+			if kl.cloud == nil {
+				// In the case where we're not using a cloud provider, the ExternalID could either have
+				// been the NodeIP, or the actual hostname.  If we switch between the two, consider it
```
Comment (ncdc) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: Do we want a link back to the PR that introduced the nodeIP issue?

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: not in the code, fine to link from this PR, which is reachable from the commit breadcrumb

**Comment by eparis** [comment]
[merge]

**Comment by ncdc** [comment]
LGTM

**Comment by smarterclayton** [comment]
LGTM

**Comment by ncdc** [comment]
```
1)  unauthenticated user should be able to log in
  - UnknownError: unknown error: [$injector:modulerr] Failed to instantiate module openshiftConsole due to:
ReferenceError: URI is not defined
```

**Comment by liggitt** [comment]
@jwforres ^

**Comment by liggitt** [comment]
[test]

**Comment by jwforres** [comment]
Opened an issue for the test problem
https://github.com/openshift/origin/issues/6651

**Comment by liggitt** [comment]
[test][merge]

**Comment by smarterclayton** [comment]
Did this ever get moved upstream?

**Comment by liggitt** [comment]
the toleration to avoid deleting/recreating? no, though I doubt upstream cares... the node just deletes itself there

**Pull request merged**